pkg/cloud/services/network/secondarycidr_test.go

```diff
@@ -0,0 +1,264 @@
+			g.Expect(err).NotTo(HaveOccurred())
+
+			if !tt.haveSecondaryCIDR {
+				mcpScope.ControlPlane.Spec.SecondaryCidrBlock = nil
+			}
+
+			s := NewService(mcpScope)
+			s.EC2Client = ec2Mock
+
+			if tt.expect != nil {
+				tt.expect(ec2Mock.EXPECT())
+			}
+
+			err = s.associateSecondaryCidr()
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+			g.Expect(err).NotTo(HaveOccurred())
+		})
+	}
+}
+
+func TestService_diassociateSecondaryCidr(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	tests := []struct {
+		name              string
+		haveSecondaryCIDR bool
+		expect            func(m *mock_ec2iface.MockEC2APIMockRecorder)
+		wantErr           bool
+	}{
+		{
+			name: "Should not disassociate secondary CIDR if no secondary cidr block info present in control plane",
+		},
+		{
+			name:              "Should return error if unable to describe VPC",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(nil, awserrors.NewFailedDependency("dependency-failure"))
+			},
+			wantErr: true,
+		},
+		{
+			name:              "Should return error if no VPC found",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(nil, nil)
+			},
+			wantErr: true,
+		},
+		{
+			name:              "Should diassociate secondary cidr block if already exist in VPC",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(&ec2.DescribeVpcsOutput{
+					Vpcs: []*ec2.Vpc{
+						{
+							CidrBlockAssociationSet: []*ec2.VpcCidrBlockAssociation{
+								{CidrBlock: aws.String("secondary-cidr")},
+							},
+						},
+					}}, nil)
+				m.DisassociateVpcCidrBlock(gomock.AssignableToTypeOf(&ec2.DisassociateVpcCidrBlockInput{})).Return(nil, nil)
+			},
+		},
+		{
+			name:              "Should return error if failed to diassociate secondary cidr block",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(&ec2.DescribeVpcsOutput{
+					Vpcs: []*ec2.Vpc{
+						{
+							CidrBlockAssociationSet: []*ec2.VpcCidrBlockAssociation{
+								{CidrBlock: aws.String("secondary-cidr")},
+							},
+						},
+					}}, nil)
+				m.DisassociateVpcCidrBlock(gomock.AssignableToTypeOf(&ec2.DisassociateVpcCidrBlockInput{})).Return(nil, awserrors.NewFailedDependency("dependency-failure"))
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			scheme, err := setupScheme()
+			g.Expect(err).NotTo(HaveOccurred())
+			cl := fake.NewClientBuilder().WithScheme(scheme).Build()
```

Comment:
Since this client is used only in `setupNewManagedControlPlaneScope` func, can we declare it inside the func?

Reply:
I have added my response in other same comment